### PR TITLE
Adding `involution` example

### PR DESCRIPTION
**Pull request by ariG23498**
In this example I have tried implementing the [involution](https://arxiv.org/abs/2103.06255) layer. The layer can be used off the shelf as it subclasses `tf.keras.layers.Layer`. 

I have also added a parallel experiment of a CNN and an INN as an image classifier. Later I go on to visualize the involution kernels and draw parallel with `self-attention` mechanism

Colab Notebook: [Involution](https://colab.research.google.com/drive/1hSIl_wMfWWzxAV4jrg85DSNYGhIueCXp?usp=sharing)

### CODE REVIEW

File: examples/vision/involution.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+"""
+
+"""
+# Introduction
+
+This example demonstrates a minimal implementation of
+[Involution: Inverting the Inherence of Convolution for Visual
+Recognition](https://arxiv.org/abs/2103.06255) by Li et. al. We will
+build the Involution Layer as a `tf.keras.layers.Layer` and try
+building an image classification model on top of it. The idea behind
```
Comment (fchollet) [inline examples/vision/involution.py:16]: Provide a description of where this model comes from, what motivated it, and why it is useful / interesting.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. There are various issues here. These are issues that aren't exclusive to this PR, so I just wrote up a contributor guide about them: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md

Please read the guide and update the example accordingly.

File: examples/vision/involution.py (hunk 1 above)
Comment (ariG23498) [reply examples/vision/involution.py]: Noted!

**Comment by ariG23498** [comment]
Hey @fchollet thank you for the guideline. I have pushed the changes according to the styleguide shared. Thank you for your feedback as well. Please let me know what you think about the recent changes that I have made.

File: examples/vision/involution.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+        # Initialize the parameters.
+        self.channel = channel
+        self.group_number = group_number
+        self.kernel_size = kernel_size
+        self.stride = stride
+        self.reduction_ratio = reduction_ratio
+
+    def build(self, input_shape):
+        # Get the shape of the input.
+        (_, H, W, C) = input_shape
```
Comment (fchollet) [inline examples/vision/involution.py:100]: Use `height`, `width`, `num_channels`

File: examples/vision/involution.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+            images=x,
+            sizes=[1, self.kernel_size, self.kernel_size, 1],
+            strides=[1, self.stride, self.stride, 1],
+            rates=[1, 1, 1, 1],
+            padding="SAME",
+        )
+
+        # Reshape the input patches to align with later operations.
+        # B, H, W, K*K, C//G, G
+        input_patches = self.input_pathces_reshape(input_patches)
```
Comment (fchollet) [inline examples/vision/involution.py:174]: "patches" (typo)

File: examples/vision/involution.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+[Convolutional Neural Network (CNN)](https://www.tensorflow.org/tutorials/images/cnn)
+tutorial from Google.
+"""
+
+"""
+## Get the CIFAR10 Dataset
+"""
+
+# Load the CIFAR10 dataset.
+print("[INFO] loading the CIFAR10 dataset...")
```
Comment (fchollet) [inline examples/vision/involution.py:233]: You can get rid of "INFO", this is not a `logging` call

File: examples/vision/involution.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+)(x)
+x = keras.layers.ReLU()(x)
+x = keras.layers.MaxPooling2D((2, 2))(x)
+x, _ = Involution(
+    channel=3, group_number=1, kernel_size=3, stride=1, reduction_ratio=2, name="inv_3"
+)(x)
+x = keras.layers.ReLU()(x)
+x = keras.layers.Flatten()(x)
+x = keras.layers.Dense(64, activation="relu")(x)
+x = keras.layers.Dense(10)(x)
```
Comment (fchollet) [inline examples/vision/involution.py:335]: Rename this `x` `outputs`

File: examples/vision/involution.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+"""
+# Visualizing Involution Kernels
+
+To visualize the kernels we take the sum of **K×K** values from each
+involution kernel. **All the representatives at different spatial
+locations frame the corresponding heat map.**
+
+The authors mention:
+
+> Our proposed involution is reminiscent of self-attention and
```
Comment (fchollet) [inline examples/vision/involution.py:418]: Just use quote characters, `>` won't work

File: examples/vision/involution.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+task, feel free to use the layer for different tasks and report your
+comparisons. 
+
+According to me the key take away of the layer is its uncanny
+relationship with that of self-attention. The intuition of spatial
+specific and channel spefic makes sense in a lot of tasks and this
+layer should be taken forward.
+
+Moving forward one can:
+- Look at [Yannick's video](https://youtu.be/pH2jZun8MoY) on
```
Comment (fchollet) [inline examples/vision/involution.py:471]: Insert a line break above (cf contributor guide)

File: examples/vision/involution.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+conv_hist = conv_model.fit(train_ds, epochs=20, validation_data=test_ds)
+
+"""
+## Involutional Neural Network
+"""
+
+# Build the involution model.
+print("[INFO] building the involution model...")
+
+inputs = keras.Input((32, 32, 3))
```
Comment (fchollet) [inline examples/vision/involution.py:318]: Specify `shape=` to be specific

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the update! It looks great. I will do a second pass with copyedits afterwards.

**Comment by ariG23498** [comment]
Took care of the requested changes 😄

**Review by fchollet — COMMENTED** [review]
Thanks for the update. I pushed copyedit changes. Please pull them and add the generated files.

**Comment by ariG23498** [comment]
@fchollet Thanks for the copyedits. I have added the generated files as asked.

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! I'll take it from here.

**Comment by fchollet** [comment]
Quick question: what did you choose to return the kernel from the layer's `call()` method? It never seems to be used.

**Comment by ariG23498** [comment]
The `kernel` was used for the visualization section. The visualizations helps in drawing a parallel between attention and involution.

**Comment by fchollet** [comment]
Got it, makes sense.

**Pull request merged**